### PR TITLE
Add coupling functors for every nonlinear system

### LINKED ISSUE #23485: MooseApp only adds relationship managers to the first nonlinear system
## Bug Description
In `MooseApp::attachRelationshipManagers`
```
    else // rm_type is algebraic or coupling
    {
      if (!_executioner && !_executor)
        mooseError("We must have an executioner by now or else we do not have to data to add "
                   "algebraic or coupling functors to in MooseApp::attachRelationshipManagers");

      // Now we've built the problem, so we can use it
      auto & problem = feProblem();
      auto & undisp_nl = problem.systemBaseNonlinear(0);
      auto & undisp_nl_dof_map = undisp_nl.dofMap();
      auto & undisp_mesh = problem.mesh().getMesh();
```
We only add managers to the first nonlinear system. 

## Steps to Reproduce
Run `moose/test/tests/misc/multiple-nl-systems/test-fv.i` with `error_on_jacobian_nonzero_reallocation=true`

## Impact
Slows down problems with improper reallocation.

### PR DESCRIPTION
**Pull request by grmnptr**
Closes #23485

### CODE REVIEW

**Review by lindsayad — CHANGES_REQUESTED** [review]
I like the coupling functor methods but not the modifications to the algebraic ghosting functor methods

**Review by lindsayad — DISMISSED** [review]

**Comment by lindsayad** [comment]
Hmm interesting...it looks like we might be able to remove those `shared_ptr` overloads. I've already approved and won't take it away, but if you want to try removing those overloads, I won't protest 😄

**Comment by grmnptr** [comment]
I believe the failures are unrelated.

File: framework/src/problems/SubProblem.C [hunk 1]
```diff
@@ -967,15 +967,13 @@ SubProblem::addAlgebraicGhostingFunctor(GhostingFunctor & algebraic_gf, bool to_
...
-  if (!n_sys)
-    return;
-
-  eq.get_system(0).get_dof_map().add_algebraic_ghosting_functor(algebraic_gf, to_mesh);
-  cloneAlgebraicGhostingFunctor(*algebraic_gf, to_mesh);
+  for (const auto nl_sys_num : make_range(numNonlinearSystems()))
+  {
+    auto & nl = systemBaseNonlinear(nl_sys_num);
+    nl.system().get_dof_map().add_coupling_functor(coupling_functor, to_mesh);
+  }
```
Comment (lindsayad) [inline framework/src/problems/SubProblem.C:976]:
```suggestion
  const auto num_nl_sys = numNonlinearSystems();
  if (!num_nl_sys)
    return;

  systemBaseNonlinear(0).system().get_dof_map().add_coupling_functor(coupling_functor, to_mesh);

  if (num_nl_sys == 1)
    return;

  auto pr = _root_coupling_functor_to_sys_clones.emplace(
      &coupling_functor, std::vector<std::shared_ptr<GhostingFunctor>>(num_nl_sys - 1));
  mooseAssert(pr.second, "We are adding a duplicate coupling functor");
  auto & clones_vec = pr.first->second;

  for (const auto i : make_range(1, num_nl_sys))
  {
    DofMap & dof_map = systemBaseNonlinear(i).system().get_dof_map();
    std::shared_ptr<GhostingFunctor> clone_coupling_functor = coupling_functor.clone();
    std::dynamic_pointer_cast<RelationshipManager>(clone_coupling_functor)
        ->init(*coupling_functor.get_mesh(), &dof_map);
    dof_map.add_coupling_functor(clone_coupling_functor, to_mesh);
    clones_vec[i - 1] = clone_coupling_functor;
  }
```
I didn't think critically enough at first. We've had problems in the past when you add a ghosting functor to a DofMap (or mesh) where the ghosting functor actually has a different DofMap (or mesh) associated with it. We have to make sure we are consistent with the ghosting-functor/DofMap (or mesh) relationship.

This change will necessitate adding a `_root_coupling_functor_to_sys_clones` data member in `SubProblem` (like we have for the algebraic ghosting functors)

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: framework/src/problems/SubProblem.C (hunk 1 above)
Comment (grmnptr) [reply framework/src/problems/SubProblem.C:976]: Got it thanks Alex!

**Review by lindsayad — APPROVED** [review]

**Pull request merged**